pkg/server/api/id_test.go

```diff
@@ -0,0 +1,171 @@
+package api_test
+
+import (
+	"testing"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/spiffe/spire/proto/spire/types"
+	"github.com/spiffe/spire/test/spiretest"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIDFromProto(t *testing.T) {
+	td := spiffeid.RequireTrustDomainFromString("domain.test")
+	workload := td.NewID("/workload")
+	reserved := td.NewID("/spire/reserved")
+	agent := td.NewID("/spire/agent/foo")
+
+	type testCase struct {
+		name        string
+		spiffeID    *types.SPIFFEID
+		expectedID  spiffeid.ID
+		expectedErr string
+	}
+
+	// These test cases are common to all of the *IDFromProto methods
+	baseCases := []testCase{
+		{
+			name:        "no SPIFFE ID",
+			expectedErr: "request must specify SPIFFE ID",
+		},
+		{
+			name:        "missing trust domain",
+			spiffeID:    &types.SPIFFEID{Path: "/workload"},
+			expectedErr: "spiffeid: trust domain is empty",
+		},
+		{
+			name:        "wrong trust domain",
+			spiffeID:    &types.SPIFFEID{TrustDomain: "otherdomain.test", Path: "/workload"},
+			expectedErr: `"spiffe://otherdomain.test/workload" is not a member of trust domain "domain.test"`,
+		},
+	}
+
+	// runTests exercises all of the test cases against the given function
+	runTests := func(t *testing.T, fn func(td spiffeid.TrustDomain, protoID *types.SPIFFEID) (spiffeid.ID, error), testCases []testCase) {
+		for _, testCase := range append(baseCases, testCases...) {
+			testCase := testCase
+			t.Run(testCase.name, func(t *testing.T) {
+				id, err := fn(td, testCase.spiffeID)
+				if testCase.expectedErr != "" {
+					require.EqualError(t, err, testCase.expectedErr)
+					return
+				}
+				require.NoError(t, err)
+				require.Equal(t, testCase.expectedID, id)
+			})
+		}
+	}
+
+	t.Run("TrustDomainMemberIDFromProto", func(t *testing.T) {
+		runTests(t, api.TrustDomainMemberIDFromProto, []testCase{
```

Comment:
We may include a case with an empty path?

Reply:
Sure! I'll add that to the base tests since it should hold true for all ID types.

Reply:
Actually there is distinct messaging, so it will be a case for each type.